### PR TITLE
thrift_proxy: simple thrift router

### PR DESCRIPTION
**Pull request by zuercher**
Provides a very basic thrift router that can route to clusters
based on method name only. A Thrift DecoderFilter interface is
introduced, but the only available filter is the Router. The
Network filter and router are capable of translating transports
and protocols but presently cannot be configured to do so.

Relates to #2247.

*Risk Level*: low
*Testing*: unit and integration testing
*Docs Changes*: protobuf documentation updated
*Release Notes*: introduced a basic thrift_proxy routing extension

Signed-off-by: Stephan Zuercher <stephan@turbinelabs.io>

### CODE REVIEW

**Comment by zuercher** [comment]
My apologies that this turned out to be such a large change. In the
end, the direction I took with the Thrift filter needed a lot of
revisions to work with a Thrift router that's capable of translating
transports and protocols. I'm willing to take a stab at breaking this
up into smaller pieces if folks think that would make reviews easier.
Just let me know.

File: api/envoy/extensions/filters/network/thrift_proxy/v2alpha1/route.proto [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+}
+
+// [#comment:next free field: 3]
+message Route {
+  // Route matching prarameters.
+  RouteMatch match = 1 [(validate.rules).message.required = true, (gogoproto.nullable) = false];
+
+  // Route request to some upstream cluster.
+  RouteAction route = 2 [(validate.rules).message.required = true, (gogoproto.nullable) = false];
+  ;
```
Comment (rgs1) [inline api/envoy/extensions/filters/network/thrift_proxy/v2alpha1/route.proto]: nit: trailing `;`

File: api/envoy/extensions/filters/network/thrift_proxy/v2alpha1/route.proto [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+}
+
+// [#comment:next free field: 2]
+message RouteMatch {
+  // If specified, the route must exactly match the request method name. As a special case, an
+  // empty string matches any request method name.
+  string method = 1;
```
Comment (rgs1) [inline api/envoy/extensions/filters/network/thrift_proxy/v2alpha1/route.proto]: is this case-sensitive?

File: source/extensions/filters/network/thrift_proxy/conn_manager.cc [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+  read_callbacks_ = &callbacks;
+}
+
+void ConnectionManager::onEvent(Network::ConnectionEvent event) {
+  if (rpcs_.empty()) {
+    return;
+  }
+
+  if (event == Network::ConnectionEvent::RemoteClose) {
+    stats_.cx_destroy_local_with_active_rq_.inc();
```
Comment (rgs1) [inline source/extensions/filters/network/thrift_proxy/conn_manager.cc]: nit: return? given we know what the event is and it can't be something else

**Comment by zuercher** [comment]
@rgs1 I think you could reuse parts of this, but not the ConnManager/Router directly. To send thrift requests to an external server (for auth or whatever), you'd likely want to extend the approach in app_exception_impl.h/.cc with some kind of code generator that would consume Thrift IDL and generate code for objects that can marshal/unmarshal themselves to/from buffers via a Protocol instance. You need some kind of AsyncThriftClient that would handle the request encoding/response decoding and transport and such. You look at the ext_authz filter for more-or-less how that would look.

File: source/extensions/filters/network/thrift_proxy/conn_manager.cc (hunk 3 above)
[the remark this answers is not in the document]
Comment (zuercher) [reply source/extensions/filters/network/thrift_proxy/conn_manager.cc]: Yes, but it also turns out that the connection pool will record these, so I'm going to remove this code.

File: source/extensions/filters/network/thrift_proxy/filters/filter.h [hunk 4]
```diff
@@ -0,0 +1,306 @@
...
+ * Decoder filter callbacks add additional callbacks.
+ */
+class DecoderFilterCallbacks {
+public:
+  virtual ~DecoderFilterCallbacks() {}
+
+  /**
+   * @return uint64_t the ID of the originating stream for logging purposes.
+   */
+  virtual uint64_t streamId() PURE;
```
Comment (brian-pane) [inline source/extensions/filters/network/thrift_proxy/filters/filter.h]: Can this be a `const` method?

Comment (brian-pane) [reply source/extensions/filters/network/thrift_proxy/filters/filter.h]: I have the same question about many of the other methods of this class.

**Comment by zuercher** [comment]
The proxy has to modify the sequence id to avoid the situation where the same sequence id from separate downstream clients is sent to a single upstream (there's an outstanding TODO in the code to actually do this). Modifying the sequence id can (for, say, the compact protocol) change the size of the message, which may necessitate modifying the transport frame.

For transports which report size (e.g. framed and header, but not unframed), I think that the mechanism for this would be:
1. the ConnManager uses the frame size to track how much more data is in the current request
2. the Router::Upstream detects (during onPoolReady) that the downstream & upstream are using the same protocol and can compute the correct transport size (handle compact proto size change and/or header size change). Some thought will have to be given as to whether this can be done in the presence of additional arbitrary DecoderFilters. Perhaps there's a way to distinguish between a filter that might modify any part of the request vs. one that only deals with headers/message begin.
3. Router::Upstream writes the modified transport start and message begin and then calls back to ConnManager to indicate that it wishes to deal purely in Buffer::Instances going forward
4. ConnManager passes remaining bytes (based on size from 1) directly to the DecoderFilter as Buffer::Instances which are written directly to the upstream connection.

File: source/extensions/filters/network/thrift_proxy/app_exception_impl.h [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+  UnknownMethod = 1,
+  InvalidMessageType = 2,
+  WrongMethodName = 3,
+  BadSequenceId = 4,
+  MissingResult = 5,
+  InternalError = 6,
+  ProtocolError = 7,
+  InvalidTransform = 8,
+  InvalidProtocol = 9,
+  UnsupportedClientType = 10,
```
Comment (fishcakez) [inline source/extensions/filters/network/thrift_proxy/app_exception_impl.h]: `fbthrift` has a few more exception types than `apache thrift`: https://github.com/facebook/fbthrift/blob/04ba53d4503d51a9a27f700b114eb7cea4678dc1/thrift/lib/cpp/TApplicationException.h#L57. Should these be supported too? As an example later in the code `InternalError` could be replaced by more specific code such as `LoadShedding`. Would that create sub-optimal experience for `apache thrift` as the enum value would not exist in the client services logic? Without using the `fbthrift` exception types there will be feature regressions for some client services as would no longer be able to differentiate.

Is `apache thrift` being treated as the standard here when there are competing standards?

Comment (zuercher) [reply source/extensions/filters/network/thrift_proxy/app_exception_impl.h]: I think short-term, we'll stick with the apache definitions. Longer term, we'll probably want to use internal identifiers for various error conditions and then provide a configuration point to control whether they get mapped to apache or fb's values.

**Pull request merged**